lib/services/unified_resource.go

```diff
@@ -0,0 +1,464 @@
+	return nil
+}
+
+// Delete deletes item by key, returns NotFound error
+// if item does not exist
+func (c *UnifiedResourceCache) delete(key []byte) error {
+	if len(key) == 0 {
+		return trace.BadParameter("missing parameter key")
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if _, ok := c.tree.Delete(&item{Key: key}); !ok {
+		return trace.NotFound("key %q is not found", string(key))
+	}
+	return nil
+}
+
+func (c *UnifiedResourceCache) getRange(startKey, endKey []byte, limit int) ([]resource, error) {
+	if len(startKey) == 0 {
+		return nil, trace.BadParameter("missing parameter startKey")
+	}
+	if len(endKey) == 0 {
+		return nil, trace.BadParameter("missing parameter endKey")
+	}
+	if limit <= 0 {
+		limit = backend.DefaultRangeLimit
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	var res []resource
+	c.tree.AscendRange(&item{Key: startKey}, &item{Key: endKey}, func(item *item) bool {
+		res = append(res, item.Value)
+		if limit > 0 && len(res) >= limit {
+			return false
+		}
+		return true
+	})
+
+	if len(res) == backend.DefaultRangeLimit {
+		c.log.Warnf("Range query hit backend limit. (this is a bug!) startKey=%q,limit=%d", startKey, backend.DefaultRangeLimit)
+	}
+	return res, nil
+}
+
+// GetUnifiedResources returns a list of all resources stored in the current unifiedResourceCollector tree
+func (c *UnifiedResourceCache) GetUnifiedResources(ctx context.Context) ([]types.ResourceWithLabels, error) {
+	if err := c.refreshStaleResources(ctx); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	result, err := c.getRange(backend.Key(prefix), backend.RangeEnd(backend.Key(prefix)), backend.NoLimit)
+	if err != nil {
+		return nil, trace.Wrap(err, "getting unified resource range")
+	}
+
+	resources := make([]types.ResourceWithLabels, 0, len(result))
+	for _, item := range result {
+		resources = append(resources, item.CloneResource())
+	}
+
+	return resources, nil
+}
+
+// ResourceGetter is an interface that provides a way to fetch all the resources
+// that can be stored in the UnifiedResourceCache
+type ResourceGetter interface {
+	NodesGetter
+	DatabaseServersGetter
+	AppServersGetter
+	WindowsDesktopGetter
+	KubernetesServerGetter
+	SAMLIdpServiceProviderGetter
+}
+
+// newWatcher starts and returns a new resource watcher for unified resources.
+func newWatcher(ctx context.Context, resourceCache *UnifiedResourceCache, cfg ResourceWatcherConfig) error {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err, "setting defaults for unified resource watcher config")
+	}
+
+	if _, err := newResourceWatcher(ctx, resourceCache, cfg); err != nil {
+		return trace.Wrap(err, "creating a new unified resource watcher")
+	}
+	return nil
+}
+
+func resourceKey(resource types.Resource) []byte {
+	return backend.Key(prefix, resource.GetName(), resource.GetKind())
+}
+
+func (c *UnifiedResourceCache) getResourcesAndUpdateCurrent(ctx context.Context) error {
+	err := c.getAndUpdateNodes(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
```

Comment:
We need to clear the previous state (i.e. delete all existing items) before inserting the new state, otherwise the cache may end up containing invalid items.  Consider the following scenario:

- Cache initializes and becomes healthy with items `A` and `B`.
- The watcher closes due to error (e.g. network interruption).
- Item `A` is deleted from the backend at and then item `C` is inserted into the backend.  The cache does not observe the events for these operations since its watcher is unhealthy.
- Cache re-establishes a healthy watcher and loads items `B` and `C` from the backend.

Without clearing first, the cache now contains items `A`, `B`, `C`, which is categorically invalid since `A` was deleted before `C` was inserted (i.e. its not just presenting an outdated/stale state... its presenting a combination of items that were never a valid state).